app/models/billboard.rb

```diff
@@ -101,6 +103,28 @@ def self.for_display(area:, user_signed_in:, user_id: nil, article: nil, user_ta
     end
   end
 
+  def self.weighted_random_selection(relation)
+    base_query = relation.to_sql
+    random_val = connection.select_value("SELECT RANDOM()").to_f
+
+    record_attrs = connection.select_all(<<-SQL).first
+      WITH base AS (#{base_query}),
+      weighted AS (
+        SELECT *, priority_weight,
+        SUM(priority_weight) OVER () AS total_weight,
+        SUM(priority_weight) OVER (ORDER BY id ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS running_weight
+        FROM base
+      )
+      SELECT *, running_weight, #{random_val} * total_weight AS random_value FROM weighted
+      WHERE running_weight >= #{random_val} * total_weight
+      ORDER BY running_weight ASC
+      LIMIT 1
+    SQL
+
+    # Instantiate an ActiveRecord object from the raw attributes
+    record_attrs ? relation.model.instantiate(record_attrs) : nil
```

Comment:
It's a little unusual to return a `nil` from a finder-type method. I maybe haven't worked on billboards recently, but what's the scenario where this query can return no results (but still have records to `sample`) ?

Reply:
I'm not 100% sure I follow this question, but both the relation we get and the further filtering (on new/priority) could result in no records returned, making the original `sample` possibly return nil, so this is basically the same logic as original.

As per the other comment, this line is now just `relation.find_by_sql([query, random_val, random_val]).first` — which is basically the same with `first` still maybe being `nil` which is logically equivalent to what it had been with `sample` unless I'm misunderstanding something.